### ISSUE TITLE
All Products block: invalid link when image has no alt text

### ISSUE THREAD

**Issue by Aljullu** [woocommerce/woocommerce-gutenberg-products-block#1654]
axe shows this error when using the _All Products_ block and at least one of the product images has no `alt` text:

> Ensures links have discernible text

Ideally we should:
* Add `aria-hidden` to the link that wraps the image when the image has no `alt` text and the product card has the `ProductTitle` block.
* Add a placeholder `alt` text when the image has no `alt` and the product card doesn't have the `ProductTitle` block.
* Do nothing when the image has an `alt` text.

**Comment by mikejolley** [comment]
@ralucaStan I checked with @Aljullu and from the original issue we agreed it's probably not feasible for atomic blocks (within the all products block) to know the properties of each other, so you can ignore his first bullet point. There are however some improvements we can make here and some related issues on first inspection:

- <del>[ ] The placeholder image needs to be aria hidden (ImagePlaceholder component)</del>
- [ ] When the placeholder is shown I see 2 images in the inspector (one is hidden) - this is a bug. It seems that the main image is rendered even when it's missing.
- [ ] Alt text appears to be missing on main image.

These issues are within the "All Products Block". When you insert this block you'll notice there is an 'edit' link which will let you setup the template and insert or remove "atomic" blocks such as product title, product link, product image etc. 

The atomic blocks themselves (where this image is rendered) are within `assets/js/atomic/blocks/product-elements`. Each has an index file which registers the block, and a block.js which handles the render.

Good luck!

**Comment by ralucaStan** [comment]
@mikejolley @Aljullu 
> The placeholder image needs to be aria hidden (ImagePlaceholder component)

I wanna propose to just use the empty alt attribute `alt=""` for the placeholder image instead of using `aria-hidden="true"`.

 This already marks it as a decorative image, [so that they can be ignored by assistive technologies, such as screen readers](https://www.w3.org/WAI/tutorials/images/decorative/).

As we want to hide the placeholder image from screenreaders it might make sense to add an aria-label to the parent anchor. 

```
<a href="#" aria-label="See product x">
    <img alt="" src="placeholder.jpg />
</a>
```

What do you think?

**Comment by mikejolley** [comment]
@ralucaStan thanks, I wasn't aware of that. Do you think including the presentation role as well would avoid this being flagged by validators? 

Found 2 resources which infer the use of both is accepted, but only in this combination:

https://fae.disability.illinois.edu/rulesets/IMAGE_1/
https://www.portent.com/blog/design-dev/decorative-images-how-to-identify-and-tag-them-for-accessibility.htm

**Comment by Aljullu** [comment]
@ralucaStan good point, setting `aria-hidden` to the image would leave the anchor without anything inside making it not accessible.

I don't think there is a perfect solution because we don't control how the merchant will set up the atomic blocks, but I wonder if we could set the `alt` attribute to placeholder images? This would avoid having to deal with ARIA and role, which is always more difficult. I mean something like this would work:

* **Placeholder images with a link:** alt set to product name
```HTML
<a href="#">
    <img alt="Product x" src="placeholder.jpg" />
</a>
```

* **Placeholder images without a link:** empty alt
```HTML
<img alt="" src="placeholder.jpg" />
```

Thoughts @ralucaStan @mikejolley?

**Comment by Aljullu** [comment]
Thinking about this, we should also handle the case where a Product image block is a link but has no alt text. Currently it results in something like:
```HTML
<a href="#">
    <img alt="" src="[IMAGE_SRC]" />
</a>
```
This is not accessible.

Should we add an extra protection and if a product image has no alt, we use the product name?

```HTML
<a href="#">
    <img alt="[PRODUCT_NAME]" src="[IMAGE_SRC]" />
</a>
```

**Comment by ralucaStan** [comment]
I am listing here the current returned code, with Case 3 bug fixed. Keep in mind, the parent could be the `<a>` or `<>`.

**1. Image not yet loaded, showing placeholder**
```js
<a href="#" rel="nofollow">
    <img alt="" hidden="" src="image.jpg" />
    <img alt="" src="placeholder.png"/>
</a>
```

**2. Image loaded, with emty alt message**
```js
<a href="" rel="nofollow">
    <img alt="" src="image.jpg" />
</a>
```

**3. No product image, showing placeholder**
```js
<a href="#" rel="nofollow">
    <img src="placeholder.png" alt="" />
</a>
```

@Aljullu @mikejolley 
 I agree, **product images** with missing alt should always use `alt="[PRODUCT_NAME]"`.

Regarding **placeholder images** I don't think they are actually decorative if I'm looking at the possible output. So, going forward I'm ok with adding `alt="[PRODUCT_NAME]"` to them as well.

What do you think?

**Comment by mikejolley** [comment]
@ralucaStan seems like the placeholder would fit inline with the example 4 on https://www.w3.org/WAI/tutorials/images/decorative/ "eye-candy". It isn't representative of the product itself, so I think alt text would be misleading.

**Comment by ralucaStan** [comment]
@mikejolley Thank could also work, like @Aljullu said there isn't here one perfect solution.

If we go with this approach than we would have to give some semantics to the link wrapping it:

```js
<a href="#" rel="nofollow" data-label="[PRODUCT_NAME]">
    <img src="placeholder.png" alt="" />
</a>
```

One question I have: can we bring the product name in this block?

**Comment by mikejolley** [comment]
@ralucaStan Yes I think we can, see `const { product } = useProductDataContext();` - this object should contain what you need.
